src/util/sockopt.c

```diff
@@ -194,3 +205,45 @@ int sockopt_get_peergroups(int fd, Log *log, uid_t uid, gid_t primary_gid, gid_t
 
         return 0;
 }
+
+int sockopt_get_peerpidfd(int fd, pid_t pid, int *ret_pidfd, bool *ret_trusted) {
+        socklen_t socklen = sizeof(int);
+        bool trusted = false;
+        int r, pidfd = -1;
+
+        assert(fd >= 0);
+        assert(ret_pidfd);
+        assert(ret_trusted);
+
+        /*
+         * New option from kernel 6.5. We'll return this with credentials,
+         * so that access policies that are resistent from PID reuse attacks
+         * can be implemented. Sets 'ret_trusted' if SO_PEERPIDFD is supported and
+         * works. If we don't have SO_PEERPIDFD, we'll fall back to the next best
+         * thing, which is to pin the pid we got from SO_PEERCRED with a pidfd.
+         * This doesn't completely close the race window, but it makes it
+         * shorter as we do this immediately after receiving the creds. In this
+         * case we set 'ret_trusted' to false, to signal to the caller that the
+         * pidfd cannot be completely trusted by the caller.
+         */
+
+        r = getsockopt(fd, SOL_SOCKET, SO_PEERPIDFD, &pidfd, &socklen);
+        if (r >= 0) {
+                if (pidfd < 0) /* API supported, but process already disappeared */
+                        return error_origin(-ESRCH);
```

Comment:
In dbus-broker, negative error codes are **always** fatal.

Any code that invokes syscalls must convert every non-fatal scenario to a proper error-domain. We **never** propagate kernel-errors. Furthermore, any of the `error_*()` calls will trigger a backtrace if called with negative kernel errors.

IOW: Only ever forward a negative errno *iff* the situation has no way to gracefully deal with them.

On a different note: The kernel returns `EINVAL` if the pid has no process-group-leader attached. `ESRCH` is returned if `SO_PEERPIDFD` is called on sockets other than UDS, which have no `sk_peer_pid`.

Reply:
So I should avoid the error_origin wrapper and just return the error directly? Or do you mean something else?

Reply:
Note that I pretty much copied the approach that other functions in the same file used, to be consistent - ie, sockopt_get_peergroups() can return error_origin(-ENOMEM), or error_origin(-errno), etc

Reply:
I've rearranged things, now they match 1:1 the other helpers, let me know if you want further changes